### PR TITLE
User Model

### PR DESCRIPTION
**Pull request by takakonishimura**
Checkout the [Engineer Workflow Doc](https://docs.google.com/presentation/d/1pdzuHG-yO7_qW6jiByaGTLztKZ6b72lBNqSdNT5WZQ8/edit) to make sure you're contributing correctly! :)

**[Tickets](https://trello.com/b/dYOQKJGv/reclass)** 
Trello Ticket: https://trello.com/c/6ABMIg4i/413-user-model
Additional Notes:

For password hashing:
- https://argon2-cffi.readthedocs.io/en/stable/argon2.html

For user authorization:
- https://dev.to/paurakhsharma/flask-rest-api-part-3-authentication-and-authorization-5935
- https://stackoverflow.com/questions/46846762/flask-jwt-extended-fake-authorization-header-during-testing-pytest
- https://stackoverflow.com/questions/62207824/typeerror-wrapper-got-an-unexpected-keyword-argument-nam-while-using-jwt-r

For parsing REST request body:
- https://marshmallow.readthedocs.io/en/stable/index.html
- https://www.cameronmacleod.com/blog/better-validation-flask-marshmallow

**[Dependencies](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.czpk4cbi7z4g)**
None

**[Explain Impact on DB](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.285lgcs4z84o)**
new tables

**[Testing](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.rs692kr58cf8)**

**New Unit Tests**
* post user - with first name (optional) ✅ , without first name (optional) ✅ , without password (required) ❌ `400`
* get username - good user_id returned ✅ , username & first name returned ✅ , fake username fails gracefully ❌ `404`
* post user login - `verified: true` for good password ✅ , `verified: false` for bad password ❌ `401`

**Ran the app locally & tested with Postman to test INSERT statements (one time activity)**
image
image

### CODE REVIEW

File: flasksrc/seed_data/users.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/*
+INSERT INTO users(id, password_hash, first_name) VALUES(################, "password_hash", "first_name");
+INSERT INTO usernames(username, toynet_userid) VALUES("username/email", ################);
+
+NOTE: the plaintext passwords here are for development login only - do NOT store production passwords in the code base!
+*/
+
+-- password: 'tayispusheen'
+INSERT INTO users(id, password_hash, first_name) VALUES(20000000000000000000, "$argon2id$v=19$m=102400,t=2,p=8$Gb99iGZpk780wPblkFzxgg$8Bnv/63zkQWkblyy/wAB+A", "Tay");
```
Comment (ajlacey) [inline flasksrc/seed_data/users.sql:9]: What hashing algorithm are we using?

File: flasksrc/schema.sql [hunk 2]
```diff
@@ -10,10 +12,20 @@ DROP TABLE IF EXISTS toynet_survey_questions;
...
+
+CREATE TABLE usernames (
+  username TEXT PRIMARY KEY,
+  toynet_userid INTEGER NOT NULL,
+  UNIQUE (toynet_userid),
+  FOREIGN KEY (toynet_userid) REFERENCES users (id)
+);
+
+CREATE TABLE users (
+  id TEXT PRIMARY KEY,
```
Comment (ajlacey) [inline flasksrc/schema.sql:25]: Why is the id text as opposed to an Integer?

File: flasksrc/__init__.py [hunk 3]
```diff
@@ -39,6 +39,11 @@ def create_app(test_config=None):
     api = Api(app)
     api.add_resource(HelloReclass, '/')
 
+    from .user import ToyNetUser, ToyNetUserByUsername, ToyNetUserLogin
+    api.add_resource(ToyNetUser, '/api/user')
+    api.add_resource(ToyNetUserByUsername, '/api/username')
+    api.add_resource(ToyNetUserLogin, '/api/login')
+
```
Comment (ajlacey) [inline flasksrc/__init__.py:46]: my ignorance here- does this just make the listed URLs available for end-user access?

File: flasksrc/schema.sql [hunk 4]
```diff
@@ -10,10 +12,20 @@ DROP TABLE IF EXISTS toynet_survey_questions;
...
 
-CREATE TABLE user (
-  id INTEGER PRIMARY KEY AUTOINCREMENT,
-  username TEXT UNIQUE NOT NULL,
-  password TEXT NOT NULL
+-- user data
+
+CREATE TABLE usernames (
+  username TEXT PRIMARY KEY,
+  toynet_userid INTEGER NOT NULL,
```
Comment (takakonishimura) [inline flasksrc/schema.sql:19]: TEXT

File: flasksrc/user.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        except ValidationError as e:
+            abort(400, message=f'malformed create user request: {e.messages}')
+
+        user_id = str(uuid.uuid1())
+        pw_hash = hasher.hash(req['password'])
+        first_name = 'first_name' in req and req['first_name'] or 'NULL'
+        username = req['username']
+
+        db = get_db()
+        try:
```
Comment (takakonishimura) [inline flasksrc/user.py:38]: sql alchemy

File: tests/test_user.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    rv = client.post('/api/user', data={
+        'username': 'bob@projectreclass.org',
+        'password': 'bobisdabomb',
+        'first_name': 'Bob',
+    })
+
+    assert rv.status_code == 200
+
+    rv_json = json.loads(rv.data.decode('utf-8'))
```
Comment (takakonishimura) [inline tests/test_user.py:35]: ?

**Pull request closed without merging**